sstables/sstables.cc

```diff
@@ -1410,6 +1410,66 @@ void sstable::write_filter() {
     write_simple<component_type::Filter>(filter_ref);
 }
 
+void sstable::maybe_rebuild_filter_from_index(uint64_t num_partitions) {
+    if (!has_component(component_type::Filter)) {
+        return;
+    }
+
+    // Skip rebuilding the bloom filter if the false positive rate based
+    // on the current bitset size is within 75% to 125% of the configured
+    // false positive rate.
+    auto curr_bitset_size = static_cast<utils::filter::bloom_filter*>(_components->filter.get())->bits().memory_size();
+    auto bitset_size_lower_bound = utils::i_filter::get_filter_size(num_partitions,
+                                                                    _schema->bloom_filter_fp_chance() * 1.25);
+    auto bitset_size_upper_bound = utils::i_filter::get_filter_size(num_partitions,
+                                                                    _schema->bloom_filter_fp_chance() * 0.75);
+    if (bitset_size_lower_bound <= curr_bitset_size && curr_bitset_size <= bitset_size_upper_bound) {
+        return;
+    }
+
+    // Consumer that adds the keys from index entries to the given bloom filter
+    class bloom_filter_builder {
+        utils::filter_ptr& _filter;
+    public:
+        bloom_filter_builder(utils::filter_ptr &filter) : _filter(filter) {}
+        void consume_entry(parsed_partition_index_entry&& e) {
+            _filter->add(to_bytes_view(e.key));
+        }
+    };
+
+    sstlog.info("Rebuilding bloom filter {}", filename(component_type::Filter));
```

Comment:
But: please add the estimated partition count and actual partition count to the log message.

Reply:
> But: please add the estimated partition count and actual partition count to the log message.

also please add the sstable origin.

Reply:
> > There is a chance for this to happen often in an overwriting workload: compacting 4 sstables with the same partition keys will estimate the sum of the partition count, but yield the maximum of the inputs. So basically we'll over-estimate by 4x. We will then spend time to rebuild the index, where we could have folded it instead. Similarly when expiring tombstones.
> > The problem is that this happens while sealing the sstable, so we're not removing backlog while this is happening.
> 
> this is less of a problem with ics since we estimate partitions per sstable (which are fixed in size) and feed that into each output one, so last output sstable is the one susceptible to overestimation. doesn't invalidate the point, but reduces the magnitude I think. also applies to lcs. also don't think it's that problematic with twcs, since expiring windows are less likely to be written (/ compacted) and usually twcs relies on full expiration.

Good point. Still, if a lot of tombstones are garbage-collected, we could end up with over-estimation.